### ISSUE TITLE
operator implementation for prod (f/w and b/w)

### ISSUE THREAD

**Issue by muthutt** [tenstorrent/tt-metal#3900]
Implementation for prod
https://pytorch.org/docs/stable/generated/torch.prod.html

Backward for prod:

Regarding the Prod backward:
Tried few simple pytorch backward program and observed that following formula
```
back(prod) = ( gradient * prod_result) /input 
```
(if the dimension is mentioned then formula is applied along dimension)
Please find the reference code in the attached images;

![image](https://github.com/tenstorrent-metal/tt-metal/assets/135061812/7f85b12c-d2ae-4ecd-9cd1-92950ed3a317)
![image](https://github.com/tenstorrent-metal/tt-metal/assets/135061812/de9c3521-09e4-4474-bc43-6b4eb96197f7)

**Comment by tt-aho** [comment]
@umadevimcw regarding your question of if we have some kind of reduce operator with multiplies.

I'm tagging @rtawfik01 and @ttmtrajkovic for this issue.

@rtawfik01 said that we don't have native support for reduce operator with multiply, but this could be done with sfpu.

Could you provide an update on requirements and priority/timeline for this?

**Comment by VirdhatchaniKN** [comment]
Sure @umadevimcw .
Hi @tt-aho . Regarding prod, we are facing an issue while multiplying the subtile values on Host (`result = result * input_buffer[input_index];`).
![Screenshot 2024-02-14 at 6 03 57 PM](https://github.com/tenstorrent-metal/tt-metal/assets/138196495/89706746-338c-4c41-8296-f6cb1327fdc8)

Could use your inputs on this.

**Comment by rtawfik01** [comment]
Just to confirm, it seems you are doing bullet point 1 & 2 on the device, and then trying bullet point 3 on host is that correct? And host is hitting a compilation error due to a missing multiply operation? I have a few questions:

1.  What is your timeline for this feature?
2. What are your perf measurements/projections? Is perf a problem for this feature?
3. Which model needs this feature?

**Comment by umadevimcw** [comment]
@rtawfik01 yes your correct 1 and 2 are on device; 3 on host
- This feature was asked as part of backward op list. Since we don't have forward prod based on discussion and information Muthu, we have started with forward implementation. ( I don't think it is for specific any model, from what we discussed with Muthu)
- Regarding perf, our initial goal is to get the functionality for the op as we don't have any reduce op for multiply

**Comment by tt-aho** [comment]
If you add the below function to the bfloat16 class in `tt_metal/common/bfloat16.hpp` this should allow you to multiply bfloat16s together.

Or just do the same functionality in your function, convert to float, multiply, then convert back to bloat16
```
    bfloat16 operator*(const bfloat16 rhs) const {
        return bfloat16(this->to_float() * rhs.to_float());
    }

```

**Comment by ruthreshx** [comment]
Hi @rtawfik01 , @ttmtrajkovic 

- We have added support for N and C for the product with the help of moreh_sum [branch: ruthresh/prod](https://github.com/tenstorrent-metal/tt-metal/pull/4933).  (Comments are yet to be address)

- With the support for H and W, we could see the bottleneck over the dims 2 and 3. In sum, they used the reduction op to reduce the dim to achieve the sum op, whereas reduction hasn't supported the mul operation yet. Any suggestions on how could we add the reduction mul from the existing reduciton ops kernel?

- Remember that this could be done using the [spfu](https://github.com/tenstorrent-metal/tt-metal/issues/3900#issuecomment-1944215377). Any idea on how this could be achievable using SPFU?

**Comment by VirdhatchaniKN** [comment]
Added Support for PROD forward op [ Draft PR [Link](https://github.com/tenstorrent-metal/tt-metal/pull/5764)]

- Support provided for all and dimensions N,C,H,W.
- Issue faced :
        - While performing prod along all dimensions, for larger shapes , the TT result is nan but the expected PyTorch result in inf. (Facing this only with GS. WH_B0 gives inf as expected)
        - Mismatch in values while performing prod along all dimensions for smaller input range - This is due to bfloat16 data type conversion
- Slow Dispatch : [Link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8080382657)
- Fast Dispatch : [Link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8080397756)

**Comment by VirdhatchaniKN** [comment]
Added Backward op support - Draft PR https://github.com/tenstorrent-metal/tt-metal/pull/5764

**Comment by VirdhatchaniKN** [comment]
Added Support for Forward and Backward PROD op - [PR Link](https://github.com/tenstorrent-metal/tt-metal/pull/5764)
Issues:

- Mismatch in values while performing prod along all dimensions for smaller input range - This is due to bfloat16 data type conversion
- While performing prod along all dimensions, for larger shapes , the TT result is nan but the expected PyTorch result is inf. (Facing this only with GS. WH_B0 gives inf as expected)

[Detailed Description](https://github.com/tenstorrent-metal/tt-metal/pull/5764#issuecomment-2052061587)

**Comment by VirdhatchaniKN** [comment]
Merged #5764 

**Description of the Issues Faced :**
List of test files:

- To test **Forward PROD** implementation along **each and all dimensions** : `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_prod.py`
- To test **Forward PROD** implementation along **all dimensions** : `tests/tt_eager/python_api_testing/unit_testing/test_prod_all.py`
- To test **Forward PROD** implementation along **N, C dimensions** alone : `tests/tt_eager/python_api_testing/unit_testing/test_prod_nc.py`
- To test **Backward PROD** implementation along **each and all dimensions** : `tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_prod.py`

Issues:

- PROD for all dimensions fails for a smaller range of input values
  - While running test `tests/tt_eager/python_api_testing/unit_testing/test_prod_all.py` below shown is the result with fill value `torch_input[:, :, :, :].fill_(1.005)`
```
torch.Size([1, 1, 32, 32])
TT Output : tensor(1632., dtype=torch.bfloat16)
Torch Output : tensor(1880., dtype=torch.bfloat16)
Out passing=False
```
```
torch.Size([1, 4, 32, 32])
TT Output : tensor(19516331393024., dtype=torch.bfloat16)
Torch Output : tensor(187904819200., dtype=torch.bfloat16)
Out passing=False
```
```
torch.Size([2, 2, 32, 32])
TT Output : tensor(19516331393024., dtype=torch.bfloat16)
Torch Output : tensor(187904819200., dtype=torch.bfloat16)
Out passing=False
```
```
torch.Size([6, 4, 32, 32]) , torch.Size([1, 3, 320, 64]), torch.Size([1, 1, 320, 320])
TT Output : tensor(inf, dtype=torch.bfloat16)
Torch Output : tensor(inf, dtype=torch.bfloat16)
Out passing=True
```
  - Without fill value, the test fails for shape `torch.Size([6, 4, 32, 32])` , `torch.Size([1, 1, 320, 320])` , `torch.Size([1, 3, 320, 64])` as the expected result is inf but the result generated in nan. Same issue when we run test file `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_prod.py` with `all_dimensions = True` (Same case as mentioned in the above point). 
  Also mentioned : [Comment Link](https://github.com/tenstorrent-metal/tt-metal/pull/5764#issuecomment-2052061587)
